packages/x-charts-pro/src/context/CartesianProviderPro/createAxisFilterMapper.ts

```diff
@@ -0,0 +1,80 @@
+import {
+  getAxisExtremum,
+  FormattedSeries,
+  ExtremumGettersConfig,
+  ZoomAxisFilter,
+  ZoomAxisFilters,
+  GetZoomAxisFilters,
+  isDefined,
+} from '@mui/x-charts/internals';
+import { ChartsAxisProps, ScaleName, AxisConfig } from '@mui/x-charts';
+import { ZoomData } from '../ZoomProvider';
+
+type CreateAxisFilterMapperParams = {
+  zoomData: ZoomData[];
+  extremumGetter: ExtremumGettersConfig;
+  formattedSeries: FormattedSeries;
+  direction: 'x' | 'y';
+};
+
+export const createAxisFilterMapper =
+  ({ zoomData, extremumGetter, formattedSeries, direction }: CreateAxisFilterMapperParams) =>
+  (axis: AxisConfig<ScaleName, any, ChartsAxisProps>, axisIndex: number): ZoomAxisFilter | null => {
+    if (typeof axis.zoom !== 'object' || axis.zoom.filterMode !== 'discard') {
+      return null;
+    }
+
+    const zoom = zoomData?.find(({ axisId }) => axisId === axis.id);
+
+    if (zoom === undefined || (zoom.start <= 0 && zoom.end >= 100)) {
+      // No zoom, or zoom with all data visible
+      return null;
+    }
+
+    let min: number;
+    let max: number;
+
+    if (axis.scaleType === 'point' || axis.scaleType === 'band') {
+      min = 0;
+      max = (axis.data?.length ?? 1) - 1;
+    } else {
+      [min, max] = getAxisExtremum(axis, extremumGetter, axisIndex === 0, formattedSeries);
+    }
+
+    const minVal = min + (zoom.start * (max - min)) / 100;
+    const maxVal = min + (zoom.end * (max - min)) / 100;
+
+    return (value, dataIndex) => {
+      const val =
+        value && typeof value === 'object' ? value[direction] : (axis.data?.[dataIndex] ?? value);
```

Comment:
@alexfauquette 
Is this what you meant? We pass `{x,y}` to the funct, the `direction` will define which we use. 

I check this before the `axis.data` part because `Scatter` ignores the axis data. (I could pass `dataIndex=undefined` too instead to always check the `axis.data` first 

```ts
axis.data?.[dataIndex] ?? (value && typeof value === 'object' ? value[direction] : value)
```

Reply:
Yes. I was thinkning about using this `{ x, y }` for all the series, even if for line and bar it means `{ x: undefined, y: value}` just for consistencies, and avoiding to get a value for which you don't know if it's x or y.

> I check this before the axis.data part because Scatter ignores the axis data.
I could pass dataIndex=undefined instead to always check the axis.data first

That's the reason why I was proposing to do a custom axis filtering in each extremums. Axes define their data, scales, .... And each series uses them to do their own logic.
But this should be ok. I don't see future charts that would require such a complex zoom feature.

About checking if a value is available first, or passing `dataIndex=undefined`, I think the first (your current) solution is the best, because:

- If you pass `undefined` it will not be feasible to support filtering on ordianl colormap
- If you verify the presence of value first, line an bar chart will not be able to provide the `x` values so they will be ok

Reply:
> Yes. I was thinkning about using this { x, y } for all the series, even if for line and bar it means { x: undefined, y: value} just for consistencies, and avoiding to get a value for which you don't know if it's x or y.

This makes the logic a bit more complex on the extremum getter side, as it needs to know which values are which, mostly for the Bar type, but I also had to set scatter dataIndex to -1

Reply:
> I also had to set scatter dataIndex to -1

Why? since the value is provided this index should not have any impact

Reply:
I thought it fixed an issue, but I was wrong.

This code produces an issue, where two points are on screen, but the axis is behaving as if only one is. I know this is caused by the `.nice`, but Idk how to fix it 🤔 

```
import * as React from 'react';
import { ScatterChartPro, ZoomData } from '@mui/x-charts-pro';

export default function AreaBaseline() {
  const [scatterZoom, setScatterZoom] = React.useState<ZoomData[]>([
    { axisId: 'b11', start: 22, end: 60 },
  ]);

  return (
    <>
      <ScatterChartPro
        xAxis={[
          { id: 'b11', data: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10], zoom: { filterMode: 'discard' } },
        ]}
        series={[
          {
            data: [
              { id: 0, x: 1, y: 100 },
              { id: 1, x: 1, y: 500 },
              { id: 2, x: 5, y: 200 },
              { id: 3, x: 10, y: 300 },
              { id: 4, x: 15, y: 400 },
              { id: 5, x: 20, y: 500 },
            ],
          },
        ]}
        zoom={scatterZoom}
        onZoomChange={(zoom) => {
          console.log(zoom);
          setScatterZoom(zoom);
        }}
        width={500}
        height={300}
      />
    </>
  );
}
```

Reply:
![Screenshot 2024-08-14 at 17 08 18](https://github.com/user-attachments/assets/62c2939e-78d4-4ec7-9d7a-13d29068e3a3)

Reply:
Because the zoom range used in the filtering does not use the `nice`.

The filtering comput it's min/max from a range of 1, 10
But the displayed axis uses `nice` so it's computed on range 0, 10


For example in the following zoom, the filtering think that the visible range is `10.17 18.10`. Which imples that the first point at x=10 as been filtered out

![image](https://github.com/user-attachments/assets/e392b4a9-7457-46e2-acb3-f8e316a5ece2)

Reply:
Yes I got that, but actually fixing it sounds quite a big change with the current implementation 🤔

Reply:
Nvm, was kinda simple change. 😆